Sources/Linq2DynamoDb.DataContext/TableDefinitionWrapper.cs

```diff
@@ -360,45 +369,120 @@ protected override void InitReader(ISupervisableEnumerable reader)
 
-            // this should never throw any exceptions
-            this.Cache.UpdateCacheAndIndexes(addedEntities, modifiedEntities, removedEntities);
-
-            return dynamoDbUpdateTask.ContinueWith
+            return Task.WhenAll(dynamoDbUpdateTasks).ContinueWith
             (
                 updateTask =>
                 {
                     if (updateTask.Exception != null)
                     {
                         // If the update failed - then removing all traces of modified entities from cache,
                         // as the update might partially succeed (in that case some of entities in cache might become stale)
-                        this.Cache.RemoveEntities(addedEntities.Keys.Union(modifiedEntities.Keys).Union(removedEntities));
+                        this.Cache.RemoveEntities(addedEntities.Keys.Union(modifiedEntities.Keys).Union(removedEntities.Keys));
 
                         throw updateTask.Exception;
                     }
+
+                    if (!string.IsNullOrEmpty(VersionPropertyName))
+                    {
+                        // Update the cache if there is a Version Property. Version properties change 
+                        // during insertion/updating so we want to have the latest values for versions 
+                        // of added and modified entities. 
+                        this.Cache.UpdateCacheAndIndexes(addedEntities, modifiedEntities, removedEntities.Keys);
+                    }
                 }
             );
         }
 
+        /// <summary>
+        /// Performs create/update operations with DynamoDB, respecting the [DynamoDBVersion] attribute if present
+        /// </summary>
+        /// <remarks>
+        /// If the entity is new (Version property has not been set) set the version property 0. If 
+        /// updating an existing entity increment the Version property by 1.
+        /// </remarks>
+        /// <param name="entitiesToAdd">Dictionary of entity keys and documents to add/update</param>
+        /// <returns>An IEnumerable containing one task for each entity to add or update.</returns>
+        private IEnumerable<Task> AddEntriesAsync(IDictionary<EntityKey, Document> entitiesToAdd) 
+        {
+            return entitiesToAdd
+                .Values
+                .Select(entry => {
+                    var putConfiguration = new PutItemOperationConfig();
+                    if (!string.IsNullOrEmpty(VersionPropertyName)) 
+                    {
+                        if (entry[VersionPropertyName] != null && 
+                            entry[VersionPropertyName].ToPrimitive(typeof(object)).Value == null)
+                        {
+                            entry[VersionPropertyName] = 0;
+
+                            putConfiguration.ConditionalExpression = new Expression() 
+                            {
+                                ExpressionStatement = $"attribute_not_exists(#key)",
```

Comment:
Andy: When performing a `PutItem` operation on an entity in DynamoDB, DynamoDB does not verify whether an object with the same key is already present. If an item with the same key is present it is overwritten. The `DataContext_AddEntity_DoesNotOverwrite_ExistingVersionedEntity` test verifies this behavior. 

Konstantin: From documentation and testing, `attribute_not_exists` is the way to go in both cases because the `attribute_not_exists` check is verifying against an object that may already be present in the table with the requested primary key (the primary key could use a hash key or a hash and range key). Since every object in a table is required to have at least a partition key, `attribute_not_exists` is checking to see if the object at the given primary key already exists in the table by checking the presence of a required attribute. My next update of the PR includes a check that demonstrates this case.

The document doesn't explicitly call this out: http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_PutItem.html

A more exacting explanation on StackOverflow: http://stackoverflow.com/questions/32833351/dynamodb-put-item-if-hash-or-hash-and-range-combination-doesnt-exist